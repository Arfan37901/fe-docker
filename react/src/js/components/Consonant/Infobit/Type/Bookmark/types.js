
import {
    bool,
    func,
    string,
} from 'prop-types';

// eslint-disable-next-line import/prefer-default-export
export const BookmarkType = {
    onClick: func,
    isBookmarked: bool,
    saveCardIcon: string,
    cardSaveText: string,
    unsaveCardIcon: string,
    cardUnsaveText: string,
    cardId: string.isRequired,
    disableBookmarkIco: bool.isRequired,
};
