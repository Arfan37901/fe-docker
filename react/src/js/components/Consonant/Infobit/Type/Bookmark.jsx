import React from 'react';
// import PropTypes from 'prop-types';

const Bookmark = () => (
    <span>Bookmark bit</span>
);

export default Bookmark;
