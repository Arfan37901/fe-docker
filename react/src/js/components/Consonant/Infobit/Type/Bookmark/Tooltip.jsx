import React from 'react';
import { string } from 'prop-types';

const TooltipType = { text: string };
const defaultProps = { text: '' };

/**
 * Tooltip (shown in 3:2 Card Footer -- primarily used with bookmark infobit)
 * Used on hover to indicate to users that they do or not do an actino
 *
 * @component
 * @example
 * const props= {
    text: String,
 * }
 * return (
 *   <Tooltip {...props}/>
 * )
 */
const Tooltip = ({ text }) => (
    <span
        className="consonant-tooltip">
        {text}
    </span>
);

Tooltip.propTypes = TooltipType;
Tooltip.defaultProps = defaultProps;

export default Tooltip;
