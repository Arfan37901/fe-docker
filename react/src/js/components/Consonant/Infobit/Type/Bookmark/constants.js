export const CLASS_PREFIX = 'consonant-bookmark-infobit';
export const STATE = {
    active: 'active',
    disabled: 'disabled',
};
