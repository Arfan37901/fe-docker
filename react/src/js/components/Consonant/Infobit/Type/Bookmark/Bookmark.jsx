import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Tooltip from './Tooltip';

function Bookmark({
    cardId,
    isBookmarked,
    saveCardIcon,
    unsaveCardIcon,
    cardSaveText,
    cardUnsaveText,
    onClick,
    disableBookmarkIco,
}) {
    const className = classNames(
        'consonant-bookmark-infobit',
        {
            'consonant-bookmark-infobit_active': isBookmarked,
            'consonant-bookmark-infobit_disabled': disableBookmarkIco,
        },
    );

    const renderIcon = () => {
        let src = '';

        if (isBookmarked && saveCardIcon) src = saveCardIcon;
        if (!isBookmarked && unsaveCardIcon) src = unsaveCardIcon;

        return (<span
            className="consonant-bookmark-infobit--ico"
            style={{ backgroundImage: src ? `url(${src})` : '' }} />);
    };

    const handleClick = (clickEvt) => {
        clickEvt.stopPropagation();
        onClick(cardId);
    };

    return (
        <button
            data-tooltip-wrapper
            type="button"
            className={className}
            onClick={handleClick}
            tabIndex="0">
            {renderIcon()}
            <Tooltip text={
                isBookmarked ? cardUnsaveText : cardSaveText
            } />
        </button>
    );
}

Bookmark.propTypes = {
    cardId: PropTypes.string.isRequired,
    saveCardIcon: PropTypes.string,
    unsaveCardIcon: PropTypes.string,
    cardSaveText: PropTypes.string,
    cardUnsaveText: PropTypes.string,
    onClick: PropTypes.func.isRequired,
    isBookmarked: PropTypes.bool.isRequired,
    disableBookmarkIco: PropTypes.bool.isRequired,
};

Bookmark.defaultProps = {
    saveCardIcon: '',
    unsaveCardIcon: '',
    cardSaveText: 'Save Card',
    cardUnsaveText: 'Unsave Card',
};

export default Bookmark;
