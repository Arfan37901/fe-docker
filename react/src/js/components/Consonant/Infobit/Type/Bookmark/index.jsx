import React, { useCallback, useMemo } from 'react';
import classNames from 'classnames';

import Tooltip from './Tooltip';
import { BookmarkType } from './types';
import { trackCardSave } from '../../../Analytics/Analytics';
import { CLASS_PREFIX, STATE, defaultProps } from './constants';

/**
 * Bookmark Infobit (shown in 3:2 Card Footer)
 * Used to save/unsave a card to local storage
 *
 * @component
 * @example
 * const props= {
    cardId: String,
    isBookmarked: String,
    saveCardIcon: String,
    unsaveCardIcon: String,
    cardSaveText: String,
    cardUnsaveText: String,
    onClick: Function,
    disableBookmarkIco: Boolean,
 * }
 * return (
 *   <Bookmark {...props}/>
 * )
 */
const Bookmark = ({
    cardId,
    onClick,
    saveCardIcon,
    cardSaveText,
    isBookmarked,
    unsaveCardIcon,
    cardUnsaveText,
    disableBookmarkIco,
}) => {
    const icon = isBookmarked ? saveCardIcon : unsaveCardIcon;
    const tooltipText = isBookmarked ? cardSaveText : cardUnsaveText;

    const className = classNames({
        CLASS_PREFIX: true,
        [`${CLASS_PREFIX}_${STATE.active}`]: isBookmarked,
        [`${CLASS_PREFIX}_${STATE.disabled}`]: disableBookmarkIco,
    });

    const handleClick = useCallback((event) => {
        event.stopPropagation();

        onClick(cardId);

        trackCardSave(cardId, !isBookmarked);
    }, [cardId, onClick, isBookmarked, trackCardSave]);

    const iconStyles = useMemo(() => ({ backgroundImage: `url(${icon})` }), [icon]);

    return (
        <button
            tabIndex="0"
            type="button"
            data-tooltip-wrapper
            className={className}
            onClick={handleClick}
            data-testid="bookmark-button">
            <span
                style={iconStyles}
                data-testid="bookmarks--ico"
                className="consonant-bookmark-infobit--ico" />
            <Tooltip text={tooltipText} />
        </button>
    );
};

Bookmark.propTypes = BookmarkType;
Bookmark.defaultProps = defaultProps;

export default Bookmark;
