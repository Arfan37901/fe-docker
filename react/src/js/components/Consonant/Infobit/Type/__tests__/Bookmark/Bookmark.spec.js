import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';

import Bookmark from '../../Bookmark/Bookmark';

const props = {
    cardId: '1',
    onClick: jest.fn(),
    disableBookmarkIco: true,
};

describe('Consonant/Bookmarks', () => {
    test('should correct render', async () => {
        const tree = renderer
            .create(<Bookmark {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
    describe('Interaction with UI', () => {
        test('should call onChange', () => {
            render(<Bookmark {...props} />);

            const boorkmarksElement = screen.getByTestId('bookmark-button');

            fireEvent.click(boorkmarksElement);

            expect(props.onClick).toBeCalled();
        });
    });
});
