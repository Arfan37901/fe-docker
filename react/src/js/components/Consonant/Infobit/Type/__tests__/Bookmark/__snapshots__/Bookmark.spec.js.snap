// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consonant/Bookmarks should correct render 1`] = `
<button
  className="consonant-bookmark-infobit consonant-bookmark-infobit_disabled"
  data-testid="bookmark-button"
  data-tooltip-wrapper={true}
  onClick={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="consonant-bookmark-infobit--ico"
    data-testid="bookmarks--ico"
    style={
      Object {
        "backgroundImage": "",
      }
    }
  />
  <span
    className="consonant-tooltip"
  >
    Save Card
  </span>
</button>
`;
